packages/next/src/server/app-render/app-render.tsx

```diff
@@ -1037,6 +1048,145 @@ async function prospectiveRuntimeServerPrerender(
+  let finishedPartial = false
+  let finishedStaletime = false
+  let passthrough = false
+
+  /**
+   * Finds the first occurrence of a subarray (`needle`) within a Uint8Array (`hay`).
+   * This is a simple byte-level search loop optimized for small, fixed ASCII patterns.
+   * It behaves like `String#indexOf` but works directly on binary data.
+   */
+  function indexOf(hay: Uint8Array, needle: Uint8Array): number {
+    const n = needle.length
+    if (hay.length < n) return -1
+    const first = needle[0] // first-byte prefilter to skip most positions quickly
+    for (let i = 0; i <= hay.length - n; i++) {
+      if (hay[i] !== first) continue
+      let j = 1
+      for (; j < n; j++) {
+        if (hay[i + j] !== needle[j]) break
+      }
+      if (j === n) return i
+    }
+    return -1
+  }
+
+  function replaceNext(
+    controller: TransformStreamDefaultController<Uint8Array>
+  ): boolean {
+    // Find whichever pattern appears first in the current buffer
+    let bestIdx = Infinity
+    let which: 'ip' | 'st' | null = null
+
+    if (!finishedPartial) {
+      const idx = indexOf(buffer, searchIP)
+      if (idx !== -1 && idx < bestIdx) {
+        bestIdx = idx
+        which = 'ip'
+      }
+    }
+    if (!finishedStaletime) {
+      const idx = indexOf(buffer, searchST)
+      if (idx !== -1 && idx < bestIdx) {
+        bestIdx = idx
+        which = 'st'
+      }
+    }
+    if (which == null) return false
+
+    const search = which === 'ip' ? searchIP : searchST
+    const repl = which === 'ip' ? replaceIP : replaceST
+
+    // flush all bytes before the match
+    if (bestIdx > 0) {
+      controller.enqueue(buffer.subarray(0, bestIdx))
+    }
+
+    // enqueue the replacement bytes
+    controller.enqueue(repl)
+
+    // keep the suffix after the matched pattern
+    buffer = buffer.subarray(bestIdx + search.length)
+
+    if (which === 'ip') {
+      finishedPartial = true
+    } else {
+      finishedStaletime = true
+    }
+
+    return true
+  }
+
+  return new TransformStream<Uint8Array, Uint8Array>({
+    transform(chunk, controller) {
+      if (passthrough) {
+        controller.enqueue(chunk)
+        return
+      }
+
+      if (buffer.length === 0) {
+        buffer = chunk
+      } else {
+        const next = new Uint8Array(buffer.length + chunk.length)
+        next.set(buffer)
+        next.set(chunk, buffer.length)
+        buffer = next
+      }
+
+      // Replace as many as possible with current bytes
+      while (!finishedPartial || !finishedStaletime) {
+        const replaced = replaceNext(controller)
+        if (!replaced) break
+      }
+
+      // If still not done, flush all but the overlap to bound memory
+      if (!finishedPartial || !finishedStaletime) {
+        if (buffer.length > keepBack) {
+          const flushLen = buffer.length - keepBack
+          controller.enqueue(buffer.subarray(0, flushLen))
+          buffer = buffer.subarray(flushLen)
+        }
```

Comment:
I would avoid copying as much as possible. We should only copy when we need to edit a chunk. I'd instead hold back one chunk from the prior transform and start searching from just before the length of the search value. Then you only need to do buffer surgery if you found your match